src/dataProvider.ts

```diff
@@ -0,0 +1,181 @@
+import * as http from 'http';
+import fetch, { RequestInit, Response, HeadersInit, BodyInit } from 'node-fetch';
+import { RateLimitError, BadRequestError, InternalServerError } from './errors';
+import delay from './delay';
+import CryptoJS from 'crypto-js';
+import Base64 from 'crypto-js/enc-base64';
+
+
+export enum DatacenterHostnames {
+    USEastCoast = 'use-data.adsrvr.org', // US East Coast
+    UKEU = 'euw-data.adsrvr.org', // UK/EU
+    APAC = 'sin-data.adsrvr.org', // APAC
+    USWestCoast = 'usw-data.adsrvr.org', // US West Coast
+    Tokyo = 'tok-data.adsrvr.org', // Tokyo
+    China = 'https://data-cn2.adsrvr.cn' // China
+}
+
+interface DataProviderOptions {
+    /**
+     * The url of the Tradedesk Data API
+     */
+    apiUrl?: string;
+
+    /**
+     * HTTP(s) Agent to use when making requests
+     */
+    agent?: http.Agent | ((parsedUrl: URL) => http.Agent);
+
+    /**
+     * Max number of retires to try a request when getting rate limited
+     */
+    maxRetries?: number
+
+    /**
+     * Max delay for an exponential back off policy for when retrying requests. Time is in seconds.
+     */
+    maxRetryDelay?: number
+
+    /**
+     * Base delay for an exponential back off policy for when retrying requests. Time is in seconds.
+     */
+    retryDelay?: number
+}
+
+class DataProvider {
+
+     /**
+     * Options used to make requests
+     */
+    public options: DataProviderOptions;
+
+    /**
+     * Provider Provisioned Secret Key
+     */
+    private secretKey: string;
+
+    constructor(secretKey: string, options: DataProviderOptions = {}) {
+        this.secretKey = secretKey
+        this.options = {
+            apiUrl: `https://${DatacenterHostnames.USEastCoast}`,
+            maxRetries: 3,
+            maxRetryDelay: 60,
+            retryDelay: 5,
+            ...options,
+        };
+    }
+
+    /**
+     * Set the API url based on an enumerated list
+     */
+    setApiUrl(dataCenter: DatacenterHostnames): DataProvider;
+
+    /**
+     * Set the API Url to an explicit url
+     */
+    setApiUrl(url: string): DataProvider;
+
+    setApiUrl(arg: any): DataProvider {
```

Comment:
Why is `any` allowed here? I would imagine that line 76 should be the implementation.

Reply:
Updated. Learned how to cast `Object.values(DatacenterHostnames)` to an `Array<string>;` to make that work.